Library/Homebrew/software_spec.rb

```diff
@@ -309,29 +309,24 @@ def initialize(formula, spec)
 
     checksum, tag, cellar = spec.checksum_for(Utils::Bottles.tag)
 
-    filename = Filename.create(formula, tag, spec.rebuild)
-
-    path, resolved_basename = spec.path_resolved_basename(@name, checksum, filename)
-
-    @resource.url("#{spec.root_url}/#{path}", select_download_strategy(spec.root_url_specs))
-    @resource.version = formula.pkg_version
-    @resource.checksum = checksum
-    @resource.downloader.resolved_basename = resolved_basename if resolved_basename.present?
     @prefix = spec.prefix
+    @tag = tag
     @cellar = cellar
     @rebuild = spec.rebuild
+
+    @resource.version = formula.pkg_version
+    @resource.checksum = checksum
+
+    root_url(spec.root_url, spec.root_url_specs)
   end
 
   def fetch(verify_download_integrity: true)
-    # add the default bottle domain as a fallback mirror
-    if @resource.download_strategy == CurlDownloadStrategy &&
-       @resource.url.start_with?(Homebrew::EnvConfig.bottle_domain)
-      fallback_url = @resource.url
-                              .sub(/^#{Regexp.escape(Homebrew::EnvConfig.bottle_domain)}/,
-                                   HOMEBREW_BOTTLE_DEFAULT_DOMAIN)
-      @resource.mirror(fallback_url) if [@resource.url, *@resource.mirrors].exclude?(fallback_url)
-    end
     @resource.fetch(verify_download_integrity: verify_download_integrity)
+  rescue DownloadError
+    raise unless fallback_on_error
+
+    fetch_tab
```

Comment:
This bit of coupling sucks but I don't know how to avoid it. `fetch_tab` will be skipped if the URL isn't a GHCR one, but if we later correct it to be then we need to go back and consider the tab.